staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go

```diff
@@ -17,330 +17,247 @@ limitations under the License.
-	return strings.HasPrefix(apiServiceName, localDelegateChainNamePrefix)
+// openAPISpecInfo is used to store OpenAPI spec with its priority.
+// It can be used to sort specs with their priorities.
+type openAPISpecInfo struct {
+	apiService v1.APIService
+	// This is the cache used to hold the specs fresh from the apiservices
+	spec cached.Replaceable[*spec.Swagger]
+
+	// The downloader is used only for non-local apiservices to
+	// re-update the spec every so often.
+	downloader cached.Data[*spec.Swagger]
 }
 
-// GetAPIServiceNames returns the names of APIServices recorded in specAggregator.openAPISpecs.
-// We use this function to pass the names of local APIServices to the controller in this package,
-// so that the controller can periodically sync the OpenAPI spec from delegation API servers.
-func (s *specAggregator) GetAPIServiceNames() []string {
-	names := make([]string, 0, len(s.openAPISpecs))
-	for key := range s.openAPISpecs {
-		names = append(names, key)
+type specAggregator struct {
+	mutex sync.Mutex
+
+	// Map of API Services' OpenAPI specs by their name
+	openAPISpecs map[string]*openAPISpecInfo
+
+	// provided for dynamic OpenAPI spec
+	openAPIVersionedService *handler.OpenAPIService
+
+	downloader *Downloader
+}
+
+func buildAndRegisterSpecAggregatorForLocalServices(downloader *Downloader, aggregatorSpec *spec.Swagger, delegationHandlers []http.Handler, pathHandler common.PathHandler) *specAggregator {
+	s := &specAggregator{
+		downloader:   downloader,
+		openAPISpecs: map[string]*openAPISpecInfo{},
+	}
+
+	cachedSpec := cached.NewStaticSource(func() cached.Result[*spec.Swagger] {
+		return cached.NewResultOK(aggregatorSpec, "never-changes")
+	})
+
+	s.addLocalSpec(fmt.Sprintf(localDelegateChainNamePattern, 0), cachedSpec)
+
+	for i, handler := range delegationHandlers {
+		name := fmt.Sprintf(localDelegateChainNamePattern, i+1)
+
+		spec := NewCacheableDownloader(downloader, handler)
+		spec = newNameDecoratedErrorCache(name, spec)
+		s.addLocalSpec(name, spec)
 	}
-	return names
+	s.openAPIVersionedService = handler.NewOpenAPIServiceLazy(s.buildMergeSpecLocked())
+	s.openAPIVersionedService.RegisterOpenAPIVersionedService("/openapi/v2", pathHandler)
+	return s
 }
 
 // BuildAndRegisterAggregator registered OpenAPI aggregator handler. This function is not thread safe as it only being called on startup.
 func BuildAndRegisterAggregator(downloader *Downloader, delegationTarget server.DelegationTarget, webServices []*restful.WebService,
 	config *common.Config, pathHandler common.PathHandler) (SpecAggregator, error) {
-	s := &specAggregator{
-		openAPISpecs: map[string]*openAPISpecInfo{},
-	}
 
-	i := 0
-	// Build Aggregator's spec
 	aggregatorOpenAPISpec, err := builder.BuildOpenAPISpecFromRoutes(restfuladapter.AdaptWebServices(webServices), config)
 	if err != nil {
 		return nil, err
 	}
+	aggregatorOpenAPISpec.Definitions = handler.PruneDefaults(aggregatorOpenAPISpec.Definitions)
```

Comment:
Question: can we build this whole thing (along with aggregatorOpenAPISpec) at the top lazily?

Reply:
Yes, but this OpenAPI is so small that I don't see the point. This also allows us to return an error to BuildAndRegisterAggregator if the aggregatorOpenAPISpec does not build, rather than waiting until the first request.